File: apps/sonar/composeApp/src/commonMain/kotlin/chat/bitchat/sonar/screens/SonarContactProfileScreen.kt
package chat.bitchat.sonar.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch
import chat.bitchat.sonar.PaySheet
import chat.bitchat.sonar.Screen
import chat.bitchat.sonar.SonarAppState
import chat.bitchat.sonar.canonicalProfileKey
import chat.bitchat.sonar.ui.SNIcon
import chat.bitchat.sonar.ui.SNIconName
import chat.bitchat.sonar.ui.SNNavHeader
import chat.bitchat.sonar.ui.SNPrimaryButton
import chat.bitchat.sonar.ui.SNSectionLabel
import chat.bitchat.sonar.ui.SNSettingsCard
import chat.bitchat.sonar.ui.SNSettingsRow
import chat.bitchat.sonar.ui.SNTone
import chat.bitchat.sonar.ui.SNTrail
import chat.bitchat.sonar.ui.SonarAvatar
import chat.bitchat.sonar.ui.SonarType
import chat.bitchat.sonar.ui.sonar

@Composable
fun SonarContactProfileScreen(state: SonarAppState, screen: Screen.ContactProfile) {
    val s = sonar
    val scope = rememberCoroutineScope()
    var showVerify by remember { mutableStateOf(false) }
    var paySheet by remember { mutableStateOf(false) }

    // Derive the peer npub from the chat members list, or accept a direct
    // npub when opened from a group member tap.
    val peerNpub = remember(screen.chatId, state.chats.size) {
        if (screen.chatId.startsWith("mesh:")) {
            val peerId = screen.chatId.removePrefix("mesh:")
            state.npubStringForPeer(peerId)?.let { canonicalProfileKey(it) }
        } else if (screen.chatId.startsWith("npub1")) {
            canonicalProfileKey(screen.chatId)
        } else {
            val chat = state.chats.firstOrNull { it.id == screen.chatId }
            val mine = canonicalProfileKey(state.npub)
            chat?.members
                ?.map { canonicalProfileKey(it) }
                ?.firstOrNull { it != mine && it.isNotBlank() }
        }
    }

    // When opened from group info with an npub (or mesh with a known npub),
    // resolve to the 1:1 DM chat id so verifyInfo/isVerified/canCall work.
    val effectiveChatId = remember(screen.chatId, peerNpub, state.chats.size) {
        val resolvedNpub = when {
            screen.chatId.startsWith("npub1") -> canonicalProfileKey(screen.chatId)
            peerNpub != null -> peerNpub
            else -> null
        }
        if (resolvedNpub != null) {
            val mine = canonicalProfileKey(state.npub)
            state.chats.firstOrNull { chat ->
                val members = chat.members.map { canonicalProfileKey(it) }
                members.size == 2 && mine in members && resolvedNpub in members
            }?.id ?: screen.chatId
        } else {
            screen.chatId
        }
    }

    val verifyInfo = remember(effectiveChatId, state.payVersion) {
        state.verifyInfo(effectiveChatId)
    }
    val verified = state.isVerified(effectiveChatId)
    val canCall = state.canCall(effectiveChatId)
    val canPay = state.hasDirectPaymentRoute(effectiveChatId)

    fun openPaySheetOrRetry() {
        scope.launch {
            val message = state.paymentDetailsUnavailableMessage(effectiveChatId)
            if (message != null) state.toast = message else paySheet = true
        }
    }

    // Find shared groups: multi-member groups where both the local user and this
    // contact are members.
    val sharedGroups = remember(state.chats.size, peerNpub) {
        if (peerNpub == null) emptyList()
        else {
            val mine = canonicalProfileKey(state.npub)
            state.chats.filter { chat ->
                state.isMultiMemberChat(chat.id) &&
                    chat.members.any { canonicalProfileKey(it) == mine } &&
                    chat.members.any { canonicalProfileKey(it) == peerNpub }
            }
        }
    }

    Column(Modifier.fillMaxSize().background(s.bg)) {
        SNNavHeader("", hairline = false, onBack = { state.back() })
        Column(
            Modifier.fillMaxSize().verticalScroll(rememberScrollState())
        ) {
            // ── Profile hero ──
            Column(
                Modifier.fillMaxWidth().padding(top = 10.dp, start = 28.dp, end = 28.dp),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                SonarAvatar(screen.name, 96.dp)
                Spacer(Modifier.height(8.dp))
                Text(
                    screen.name,
                    color = s.text,
                    fontSize = 24.sp,
                    fontWeight = FontWeight.Black
                )
                Spacer(Modifier.height(6.dp))
                Box(
                    Modifier.clip(CircleShape).background(s.surface2)
                        .padding(horizontal = 11.dp, vertical = 4.dp)
                ) {
                    Text(
                        shortKey(peerNpub),
                        color = s.text3,
                        style = SonarType.mono(12.0)
                    )
                }
            }

            Spacer(Modifier.height(18.dp))

            // ── Action buttons row ──
            Row(
                Modifier.fillMaxWidth().padding(horizontal = 28.dp),
                horizontalArrangement = Arrangement.Center
            ) {
                ActionCircle(
                    icon = SNIconName.Lock,
                    label = "Message",
                    onClick = {
                        if (effectiveChatId != screen.chatId) {
                            val dmChat = state.chats.firstOrNull { it.id == effectiveChatId }
                            if (dmChat != null) state.openChat(dmChat)
                            else state.back()
                        } else {
                            state.back()
                        }
                    }
                )
                Spacer(Modifier.width(28.dp))
                ActionCircle(
                    icon = SNIconName.Phone,
                    label = "Call",
                    enabled = canCall,
                    onClick = {
                        if (canCall) state.placeCall(effectiveChatId, screen.name, false)
                        else state.toast = "No call route to this peer yet."
                    }
                )
                Spacer(Modifier.width(28.dp))
                ActionCircle(
                    icon = SNIconName.Coin,
                    label = "Pay",
                    enabled = canPay,
                    onClick = { openPaySheetOrRetry() }
                )
                Spacer(Modifier.width(28.dp))
                ActionCircle(
                    icon = if (verified) SNIconName.ShieldCheck else SNIconName.Shield,
                    label = if (verified) "Verified" else "Verify",
                    onClick = { showVerify = !showVerify }
                )
            }

            Spacer(Modifier.height(20.dp))

            // ── Verify inline section ──
            if (showVerify) {
                VerifyInline(
                    peerName = screen.name,
                    myName = state.nick.ifBlank { "you" },
                    info = verifyInfo,
                    onVerify = { state.markVerified(effectiveChatId) },
                    onDismiss = { showVerify = false }
                )
                Spacer(Modifier.height(12.dp))
            }

            // ── Identity card ──
            SNSectionLabel("Identity")
            SNSettingsCard {
                SNSettingsRow(
                    icon = SNIconName.Key,
                    tone = SNTone.Cyan,
                    label = "Public key",
                    value = shortKey(peerNpub),
                    valueMono = true,
                    trail = SNTrail.None
                ) {}
                SNSettingsRow(
                    icon = SNIconName.Lock,
                    label = "Key fingerprint",
                    value = verifyInfo.safety.firstOrNull()?.take(10)?.let { "$it..." } ?: "n/a",
                    valueMono = true,
                    trail = SNTrail.None,
                    divider = false
                ) {}
            }

            // ── Safety card ──
            SNSectionLabel("Safety")
            SNSettingsCard {
                SNSettingsRow(
                    icon = if (verified) SNIconName.ShieldCheck else SNIconName.Shield,
                    tone = if (verified) SNTone.Cyan else SNTone.Default,
                    label = "Safety number",
                    value = if (verified) "Verified" else "Not verified",
                    trail = SNTrail.Chevron
                ) { showVerify = !showVerify }
                SNSettingsRow(
                    icon = SNIconName.Lock,
                    tone = SNTone.Cyan,
                    label = "End-to-end encrypted",
                    sub = "Messages are encrypted with the Signal protocol",
                    trail = SNTrail.None,
                    divider = false
                ) {}
            }

            // ── Shared groups ──
            SNSectionLabel("Shared groups")
            if (sharedGroups.isEmpty()) {
                Text(
                    "No shared groups",
                    color = s.text3,
                    fontSize = 14.sp,
                    modifier = Modifier.padding(start = 18.dp, bottom = 8.dp)
                )
            } else {
                SNSettingsCard {
                    sharedGroups.forEachIndexed { idx, group ->
                        val groupName = state.chatTitle(group)
                        Row(
                            Modifier.fillMaxWidth()
                                .clickable { state.openChat(group) }
                                .padding(horizontal = 14.dp, vertical = 11.dp),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            SonarAvatar(groupName, 36.dp)
                            Spacer(Modifier.width(12.dp))
                            Column(Modifier.weight(1f)) {
                                Text(
                                    groupName,
                                    color = s.text,
                                    fontSize = 16.sp,
                                    fontWeight = FontWeight.Medium
                                )
                                Text(
                                    "${group.members.size} members",
                                    color = s.text3,
                                    fontSize = 12.5.sp
                                )
                            }
                        }
                        if (idx < sharedGroups.lastIndex) {
                            Box(
                                Modifier.fillMaxWidth().padding(start = 60.dp)
                                    .height(1.dp).background(s.hairline)
                            )
                        }
                    }
                }
            }

            // ── Actions card ──
            SNSectionLabel("Actions")
            SNSettingsCard {
                SNSettingsRow(
                    icon = SNIconName.X,
                    tone = SNTone.Red,
                    label = "Block contact",
                    danger = true,
                    trail = SNTrail.None
                ) { state.toast = "Coming soon" }
                SNSettingsRow(
                    icon = SNIconName.Trash,
                    tone = SNTone.Red,
                    label = "Delete chat",
                    danger = true,
                    trail = SNTrail.None,
                    divider = false
                ) { state.toast = "Coming soon" }
            }

            Spacer(Modifier.height(40.dp))
        }
    }
    if (paySheet && canPay) {
        val isMesh = effectiveChatId.startsWith("mesh:")
        PaySheet(
            peerName = screen.name,
            balanceSats = state.walletBalanceSats(),
            mesh = isMesh,
            fiatOf = { state.fiatOrNull(it) },
            onSend = { sats ->
                scope.launch { state.sendPay(effectiveChatId, sats)?.let { state.toast = it } }
            },
            onClose = { paySheet = false }
        )
    }
}

/** Circular action button with icon and label below. */
@Composable
private fun ActionCircle(
    icon: SNIconName,
    label: String,
    enabled: Boolean = true,
    onClick: () -> Unit,
) {
    val s = sonar
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Box(
            Modifier.size(52.dp)
                .clip(CircleShape)
                .background(if (enabled) s.accentSoft else s.surface2)
                .clickable(enabled = enabled, onClick = onClick),
            contentAlignment = Alignment.Center
        ) {
            SNIcon(
                icon, 22.dp,
                if (enabled) s.accentDeep else s.text3,
                weight = 2.1f
            )
        }
        Spacer(Modifier.height(5.dp))
        Text(
            label,
            color = s.text2,
            fontSize = 11.sp,
            fontWeight = FontWeight.Medium
        )
    }
}

/** Inline verify section — shows safety numbers and verify button. */
@Composable
private fun VerifyInline(
    peerName: String,
    myName: String,
    info: chat.bitchat.sonar.SonarVerify,
    onVerify: () -> Unit,
    onDismiss: () -> Unit,
) {
    val s = sonar
    Column(
        Modifier.fillMaxWidth().padding(horizontal = 14.dp)
            .clip(RoundedCornerShape(18.dp)).background(s.surface)
            .padding(20.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(
            "Verify safety numbers",
            color = s.text,
            fontSize = 18.sp,
            fontWeight = FontWeight.Bold
        )
        Spacer(Modifier.height(14.dp))
        Row(horizontalArrangement = Arrangement.spacedBy(28.dp)) {
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                SonarAvatar(myName, 48.dp, presence = false)
                Spacer(Modifier.height(4.dp))
                Text(myName, color = s.text2, fontSize = 12.sp)
            }
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                SonarAvatar(peerName, 48.dp, presence = false)
                Spacer(Modifier.height(4.dp))
                Text(peerName, color = s.text2, fontSize = 12.sp)
            }
        }
        Spacer(Modifier.height(16.dp))
        if (info.safety.isEmpty()) {
            Text(
                info.note ?: "Safety numbers aren't available yet.",
                color = s.text2,
                fontSize = 13.5.sp,
                textAlign = TextAlign.Center
            )
        } else {
            Text(
                "Compare these numbers with $peerName in person or on a call. If they match, this chat is end-to-end encrypted and nobody is in the middle.",
                color = s.text2,
                fontSize = 13.5.sp,
                lineHeight = 18.sp,
                textAlign = TextAlign.Center
            )
            Spacer(Modifier.height(14.dp))
            // 3 rows x 4 groups, monospace.
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                listOf(0, 4, 8).forEach { row ->
                    Text(
                        info.safety.subList(row, row + 4).joinToString(" "),
                        color = s.text,
                        style = SonarType.mono(15.0),
                        modifier = Modifier.padding(vertical = 3.dp)
                    )
                }
            }
            Spacer(Modifier.height(18.dp))
            if (info.verified) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    SNIcon(SNIconName.ShieldCheck, 16.dp, s.green)
                    Spacer(Modifier.width(6.dp))
                    Text(
                        "Verified",
                        color = s.green,
                        fontSize = 15.sp,
                        fontWeight = FontWeight.Bold
                    )
                }
            } else {
                SNPrimaryButton("They match — mark as verified") { onVerify() }
            }
        }
        Spacer(Modifier.height(8.dp))
        Box(
            Modifier.fillMaxWidth().height(44.dp).clickable(onClick = onDismiss),
            contentAlignment = Alignment.Center
        ) {
            Text(
                "Close",
                color = s.text2,
                fontSize = 15.sp,
                fontWeight = FontWeight.SemiBold
            )
        }
    }
}

File: ios/bitchat/Views/Sonar/SonarContactProfileScreen.swift
//
// SonarContactProfileScreen.swift
// bitchat
//
// Contact profile screen: hero section, action buttons (message/call/verify),
// identity & safety cards, shared groups, block/delete. Ported from the
// Compose Multiplatform SonarContactProfileScreen.
//
// This is free and unencumbered software released into the public domain.
// For more information, see <https://unlicense.org>
//

import SwiftUI
#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

struct SonarContactProfileScreen: View {
    @EnvironmentObject private var store: SonarAppStore
    let peerId: String
    let peerName: String

    @State private var showVerify = false
    @State private var paySheet = false
    @State private var walletSheet = false
    @State private var toast: String?

    private var effectiveChatId: String {
        guard peerId.hasPrefix("npub1") else { return peerId }
        if let group = store.marmotGroup(forNpub: peerId) {
            return SonarAppStore.marmotIDPrefix + group.id
        }
        if let peerKey = store.sonarPeerKey(forNpub: peerId) {
            return peerKey
        }
        return peerId
    }

    private var resolvedNpub: String {
        if peerId.hasPrefix("npub1") { return peerId }
        if let npub = store.sonarProfile(peerId)?.npub, !npub.isEmpty {
            return npub
        }
        if let group = store.marmotGroup(forConversationId: peerId),
           store.marmot.isDirectGroup(group),
           let otherNpub = store.marmot.otherMembers(in: group).first {
            return otherNpub
        }
        return ""
    }

    private var verified: Bool { store.isVerified(effectiveChatId) }
    private var info: SNVerifyInfo { store.verifyInfo(for: effectiveChatId) }
    private var canPay: Bool { store.paymentCapable(effectiveChatId) }
    private var walletReady: Bool {
        if case .ready = store.walletState { return true }
        return false
    }

    private var peerFullKey: String {
        let npub = resolvedNpub
        return npub.isEmpty ? (store.sonarProfile(peerId)?.npub ?? "") : npub
    }

    private var peerShortKey: String {
        let npub = peerFullKey
        guard npub.count > 16 else { return "n/a" }
        return "\(npub.prefix(10))…\(npub.suffix(4))"
    }

    private var sharedGroups: [SNDMRow] {
        let npub = resolvedNpub
        guard !npub.isEmpty else { return [] }
        return store.dmRows.filter { row in
            guard row.isMarmot, store.isMultiMemberMarmotGroupId(row.id) else { return false }
            let members = store.groupMemberContacts(forConversationId: row.id)
            return members.contains { $0.npub == npub }
        }
    }

    var body: some View {
        VStack(spacing: 0) {
            SNNavHeader(hairline: false, onBack: { store.pop() }) {
                EmptyView()
            }

            ScrollView {
                VStack(spacing: 0) {
                    // ── Hero ──
                    VStack(spacing: 0) {
                        SonarAvatar(name: peerName, size: 96)
                        Text(verbatim: peerName)
                            .font(SonarTheme.uiFont(size: 24, weight: .black))
                            .foregroundColor(SonarTheme.text)
                            .padding(.top, 8)
                        Text(verbatim: peerShortKey)
                            .font(SonarTheme.monoFont(size: 12))
                            .foregroundColor(SonarTheme.text3)
                            .padding(.horizontal, 11)
                            .padding(.vertical, 4)
                            .background(Capsule().fill(SonarTheme.surface2))
                            .padding(.top, 6)
                    }
                    .frame(maxWidth: .infinity)
                    .padding(.top, 10)
                    .padding(.horizontal, 28)

                    // ── Action buttons ──
                    HStack(spacing: 28) {
                        profileAction(icon: .lock, label: "Message") {
                            if effectiveChatId != peerId {
                                store.push(.dm(effectiveChatId))
                            } else {
                                store.pop()
                            }
                        }
                        profileAction(icon: .phone, label: "Call", enabled: store.canCall(effectiveChatId)) {
                            if store.canCall(effectiveChatId) {
                                store.placeCall(effectiveChatId, video: false)
                            }
                        }
                        profileAction(icon: .coin, label: "Pay", enabled: canPay) {
                            openPaySheetOrWallet()
                        }
                        profileAction(
                            icon: verified ? .shieldCheck : .shield,
                            label: verified ? "Verified" : "Verify"
                        ) {
                            showVerify.toggle()
                        }
                    }
                    .padding(.top, 18)
                    .padding(.bottom, 20)

                    // ── Verify inline ──
                    if showVerify {
                        verifySection
                            .padding(.bottom, 12)
                    }

                    // ── Identity ──
                    SNSectionLabel("Identity")
                    SNSettingsCard {
                        SNSettingsRow(
                            icon: .key, tone: .cyan, label: "Public key",
                            value: peerShortKey, valueMono: true
                        ) {
                            copyKey()
                        }
                        SNSettingsRow(
                            icon: .lock, label: "Key fingerprint",
                            value: info.safety.first.map { "\($0.prefix(10))…" } ?? "n/a",
                            valueMono: true, trail: .none, divider: false
                        ) {}
                    }

                    // ── Safety ──
                    SNSectionLabel("Safety")
                    SNSettingsCard {
                        SNSettingsRow(
                            icon: verified ? .shieldCheck : .shield,
                            tone: verified ? .cyan : .neutral,
                            label: "Safety number",
                            value: verified ? "Verified" : "Not verified"
                        ) { showVerify.toggle() }
                        SNSettingsRow(
                            icon: .lock, tone: .cyan,
                            label: "End-to-end encrypted",
                            sub: "Messages are encrypted with the Signal protocol",
                            trail: .none, divider: false
                        ) {}
                    }

                    // ── Shared groups ──
                    SNSectionLabel("Shared groups")
                    if sharedGroups.isEmpty {
                        Text("No shared groups")
                            .font(SonarTheme.uiFont(size: 14))
                            .foregroundColor(SonarTheme.text3)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .padding(.leading, 18)
                            .padding(.bottom, 8)
                    } else {
                        SNSettingsCard {
                            ForEach(Array(sharedGroups.enumerated()), id: \.element.id) { i, group in
                                Button {
                                    store.push(.dm(group.id))
                                } label: {
                                    HStack(spacing: 12) {
                                        SonarAvatar(name: group.title, size: 36)
                                        VStack(alignment: .leading, spacing: 2) {
                                            Text(verbatim: group.title)
                                                .font(SonarTheme.uiFont(size: 16, weight: .medium))
                                                .foregroundColor(SonarTheme.text)
                                            Text(verbatim: "Group")
                                                .font(SonarTheme.uiFont(size: 12.5))
                                                .foregroundColor(SonarTheme.text3)
                                        }
                                        .frame(maxWidth: .infinity, alignment: .leading)
                                    }
                                    .padding(.horizontal, 14)
                                    .padding(.vertical, 11)
                                    .contentShape(Rectangle())
                                    .overlay(alignment: .bottom) {
                                        if i < sharedGroups.count - 1 {
                                            Rectangle().fill(SonarTheme.hairline).frame(height: 1).padding(.leading, 60)
                                        }
                                    }
                                }
                                .buttonStyle(SNRowPressStyle())
                            }
                        }
                    }

                    // ── Actions ──
                    SNSectionLabel("Actions")
                    SNSettingsCard {
                        SNSettingsRow(
                            icon: .x, tone: .red, label: "Block contact",
                            danger: true, trail: .none
                        ) { showToast("Coming soon") }
                        SNSettingsRow(
                            icon: .trash, tone: .red, label: "Delete chat",
                            danger: true, trail: .none, divider: false
                        ) { showToast("Coming soon") }
                    }

                    Color.clear.frame(height: 40)
                }
            }
        }
        .snSheet(isPresented: $paySheet, title: "Send money · \(peerName)") {
            SNPaySheet(
                peerName: peerName,
                balance: store.balanceSats ?? 0,
                transport: store.dmTransport(effectiveChatId),
                money: { store.money($0) },
                fiatText: { store.fiatText($0) },
                onClose: { paySheet = false },
                onSend: { sats in
                    Task {
                        if let message = await store.sendPay(effectiveChatId, sats: sats) {
                            showToast(message)
                        }
                    }
                }
            )
        }
        .snSheet(isPresented: $walletSheet, title: "Your wallet") {
            SNWalletSheetContent(onClose: { walletSheet = false })
        }
        .background(SonarTheme.bg.ignoresSafeArea())
        .overlay(alignment: .bottom) {
            if let toast {
                Text(verbatim: toast)
                    .font(SonarTheme.uiFont(size: 13.5, weight: .medium))
                    .foregroundColor(SonarTheme.text)
                    .padding(EdgeInsets(top: 11, leading: 16, bottom: 11, trailing: 16))
                    .background(
                        RoundedRectangle(cornerRadius: 13, style: .continuous)
                            .fill(SonarTheme.surface2)
                            .shadow(color: Color.black.opacity(0.18), radius: 12, y: 6)
                    )
                    .padding(.horizontal, 24)
                    .padding(.bottom, 88)
                    .transition(.move(edge: .bottom).combined(with: .opacity))
            }
        }
        .animation(.easeOut(duration: 0.2), value: toast)
    }

    private func copyKey() {
        let key = peerFullKey
        guard !key.isEmpty else { return }
        #if canImport(UIKit)
        UIPasteboard.general.string = key
        #elseif canImport(AppKit)
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(key, forType: .string)
        #endif
        showToast("Public key copied")
    }

    private func showToast(_ text: String) {
        toast = text
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.6) {
            if toast == text { toast = nil }
        }
    }

    private func openPaySheetOrWallet() {
        guard walletReady else {
            walletSheet = true
            return
        }
        Task {
            if let message = await store.paymentDetailsUnavailableMessage(effectiveChatId) {
                showToast(message)
                return
            }
            paySheet = true
        }
    }

    private func profileAction(icon: SNIconName, label: String, enabled: Bool = true, action: @escaping () -> Void) -> some View {
        VStack(spacing: 5) {
            Button(action: action) {
                Circle()
                    .fill(enabled ? SonarTheme.accentSoft : SonarTheme.surface2)
                    .frame(width: 52, height: 52)
                    .overlay(
                        SNIcon(name: icon, size: 22, weight: 2.1)
                            .foregroundColor(enabled ? SonarTheme.accentDeep : SonarTheme.text3)
                    )
            }
            .buttonStyle(SNScaleStyle(scale: 0.94))
            .disabled(!enabled)
            Text(verbatim: label)
                .font(SonarTheme.uiFont(size: 11, weight: .medium))
                .foregroundColor(SonarTheme.text2)
        }
    }

    @ViewBuilder
    private var verifySection: some View {
        VStack(spacing: 0) {
            Text("Verify safety numbers")
                .font(SonarTheme.uiFont(size: 18, weight: .bold))
                .foregroundColor(SonarTheme.text)
                .padding(.bottom, 14)

            HStack(spacing: 28) {
                VStack(spacing: 4) {
                    SonarAvatar(name: store.nick.isEmpty ? "you" : store.nick, size: 48)
                    Text(verbatim: store.nick.isEmpty ? "you" : store.nick)
                        .font(SonarTheme.uiFont(size: 12))
                        .foregroundColor(SonarTheme.text2)
                }
                VStack(spacing: 4) {
                    SonarAvatar(name: peerName, size: 48)
                    Text(verbatim: peerName)
                        .font(SonarTheme.uiFont(size: 12))
                        .foregroundColor(SonarTheme.text2)
                }
            }
            .padding(.bottom, 16)

            if info.available {
                Text("Compare these numbers with \(peerName) in person or on a call. If they match, this chat is end-to-end encrypted and nobody is in the middle.")
                    .font(SonarTheme.uiFont(size: 13.5))
                    .lineSpacing(13.5 * 0.3)
                    .foregroundColor(SonarTheme.text2)
                    .multilineTextAlignment(.center)
                    .padding(.bottom, 14)

                VStack(spacing: 0) {
                    ForEach([0, 4, 8], id: \.self) { row in
                        Text(verbatim: info.safety[row..<(row + 4)].joined(separator: "\u{2002}"))
                            .font(SonarTheme.monoFont(size: 15))
                            .kerning(15 * 0.06)
                            .foregroundColor(SonarTheme.text)
                            .frame(height: 15 * 2)
                    }
                }
                .padding(.bottom, 18)

                if verified {
                    HStack(spacing: 6) {
                        SNIcon(name: .shieldCheck, size: 16)
                            .foregroundColor(SonarTheme.green)
                        Text("Verified")
                            .font(SonarTheme.uiFont(size: 15, weight: .bold))
                            .foregroundColor(SonarTheme.green)
                    }
                } else {
                    SNPrimaryButton(label: "They match — mark as verified") {
                        store.markVerified(effectiveChatId)
                    }
                    .padding(.horizontal, 8)
                }
            } else {
                Text(verbatim: info.note ?? "Safety numbers aren\u{2019}t available yet.")
                    .font(SonarTheme.uiFont(size: 13.5))
                    .lineSpacing(13.5 * 0.3)
                    .foregroundColor(SonarTheme.text2)
                    .multilineTextAlignment(.center)
            }

            Button { showVerify = false } label: {
                Text("Close")
                    .font(SonarTheme.uiFont(size: 15, weight: .semibold))
                    .foregroundColor(SonarTheme.text2)
                    .frame(maxWidth: .infinity)
                    .frame(height: 44)
                    .contentShape(Rectangle())
            }
            .buttonStyle(SNRowPressStyle(cornerRadius: 12))
            .padding(.top, 8)
        }
        .padding(20)
        .background(
            RoundedRectangle(cornerRadius: 18, style: .continuous)
                .fill(SonarTheme.surface)
        )
        .padding(.horizontal, 14)
    }
}
